eth-connector/src/lib.rs

```diff
@@ -556,31 +677,52 @@ impl FundsFinish for EthConnectorContract {
 
         log!("Finish deposit with the amount: {}", deposit_call.amount);
 
-        // Mint - calculate new balances
-        self.mint_eth_on_near(&deposit_call.new_owner_id, deposit_call.amount);
         // Store proof only after `mint` calculations
         self.record_proof(&deposit_call.proof_key).sdk_unwrap();
 
-        // Mint tokens to recipient minus fee
-        deposit_call.msg.map_or_else(
-            || PromiseOrValue::Value(None),
-            |msg| {
-                let args = TransferCallCallArgs::try_from_slice(&msg)
-                    .map_err(|_| crate::errors::ERR_BORSH_DESERIALIZE)
-                    .sdk_unwrap();
-                let promise = self.internal_ft_transfer_call(
-                    env::predecessor_account_id(),
-                    args.receiver_id,
-                    args.amount.into(),
-                    args.memo,
-                    args.msg,
-                );
-                match promise {
-                    PromiseOrValue::Promise(p) => PromiseOrValue::Promise(p),
-                    PromiseOrValue::Value(v) => PromiseOrValue::Value(Some(v)),
-                }
-            },
-        )
+        if let Some(msg) = deposit_call.msg {
+            // Mint - calculate new balances
+            self.mint_eth_on_near(&env::current_account_id(), deposit_call.amount);
+
+            // Transfer tokens to recipient minus fee
+            let args = TransferCallCallArgs::try_from_slice(&msg)
+                .map_err(|_| crate::errors::ERR_BORSH_DESERIALIZE)
+                .sdk_unwrap();
+
+            let fee_amount = self.calculate_fee_amount(
+                deposit_call.amount.into(),
+                FeeType::Deposit,
+                Some(args.receiver_id.clone()),
+            );
+            let amount_to_transfer = deposit_call.amount.checked_sub(fee_amount.0).sdk_unwrap();
+            // Early return if the fee is higher than the transferred amount
+            if amount_to_transfer == 0 && fee_amount.0 != 0 {
+                return PromiseOrValue::Value(Some(U128(0)));
+            }
+
+            let promise = self.internal_ft_transfer_call(
+                env::current_account_id(),
+                args.receiver_id,
+                amount_to_transfer.into(),
+                args.memo,
+                args.msg,
+            );
+            match promise {
+                PromiseOrValue::Promise(p) => PromiseOrValue::Promise(p),
+                PromiseOrValue::Value(v) => PromiseOrValue::Value(Some(v)),
+            }
+        } else {
+            let fee_amount =
+                self.calculate_fee_amount(deposit_call.amount.into(), FeeType::Deposit, None);
+            let amount_to_transfer = deposit_call.amount.checked_sub(fee_amount.0).sdk_unwrap();
+
```

Comment:
In this case, it's rather `deposit_amount` as we don't create any transfer per se.

Reply:
10d1ab3ac41e26e55b890e87629fee46333d37aa